ENH: Added Yeo-Johnson power transformation

NicolasHug opened the pull request:
Closes  #6141.

This PR adds support for the Yeo-Johnson power transform. Unlike the (already implemented) Box-Cox transform, Yeo-Johnson is able to deal with negative values.

I based the implementation on that of scikit-learn (scikit-learn/scikit-learn#11520) and tried to mirror that of Box-Cox.

Done:
- [X] `yeojohnson`
- [X] `yeojohnson_llf` (log likelihood function)
- [X] `yeojohnson_normmax` (optimization function, using the brent optimizer just like boxcox)
- [X] `yeojohnson_normplot`
- [X] tests
- [X] docstrings
- [X] versionadded tag
- [X] updated release notes
- [X] updated Thanks.txt


Differences with the boxcox implementation:
- Here the code is pure numpy. The actual boxcox transformation is Cythonized
- Only MLE is available to optimize lambda parameter. Boxcox also supports `pearsonr`.
- `yeojohnson` does not supports the `alpha` parameter for returning a confidence interval.
- ~~`yeojohnson_llf` only supports 1D arrays (same for 'yeojohnson' and 'boxcox' anyway)~~ `yeojohnson_llf` now behaves like `boxcox_llf`

chrisb83 commented:
Thanks, this looks like a very nice feature. I will try to review it soon

chrisb83 commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    pos = x >= 0  # binary mask
+
+    # when x >= 0
+    if abs(lmbda) < 1e-19:
```
why `1e-19`?

ilayn replied:
That won't make a strict rule if regular floats are concerned. You can use `np.spacing(1.)` instead.

chrisb83 commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+
+    # when x >= 0
+    if abs(lmbda) < 1e-19:
+        out[pos] = np.log(x[pos] + 1)
```
use `log1p` for better accuracy for small x, same in line 1363

chrisb83 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1200,6 +1200,88 @@ def test_empty(self):
+
+class TestYeojohnson(object):
+
+    def test_fixed_lmbda(self):
```
can you add further test cases for negative x and `lmbda == 2` / `lmbda != 2`?

and also vector x with negative and positive numbers?

NicolasHug replied:
Just did in 1c898bb

chrisb83 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1200,6 +1200,88 @@ def test_empty(self):
+
+    def test_mle(self):
+        maxlog = stats.yeojohnson_normmax(self.x)
+        assert_allclose(maxlog, 1.876393, rtol=1e-6)
```
why is this the expected result?

chrisb83 commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+        llf = N/2 \log(\sigma^2) + (\lambda - 1)
+              \sum_i \text{ sign }(x_i)\log(|x_i| + 1)
+
+    where :math:`\sigma^2` is estimated variance of the the Yeo-Johnson
```
typically denoted by sigma hat, but not sure if this is more clear

chrisb83 commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    trans = _yeojohnson_transform(data, lmb)
+
+    # Estimated mean and variance of the normal distribution
+    est_mean = trans.sum() / n_samples
```
slightly shorter: `trans.mean()` which can be inserted directly into the line below
or use x.var() directly

chrisb83 commented:
regarding Tests: the original paper of Yeo and Johnson contains an example at the end

NicolasHug commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    We now use `yeojohnson` to transform the data so it's closest to normal:
+
+    >>> ax2 = fig.add_subplot(212)
+    >>> xt, _ = stats.yeojohnson(x)
```
You mean to add a `print` statement?

NicolasHug commented:
Thanks for the review I think I've addressed all comments

ilayn commented on doc/release/1.2.0-notes.rst:
```diff
@@ -47,9 +55,12 @@ Other changes
 Issues closed
 -------------
 
+* `#6141 <https://github.com/scipy/scipy/issues/6141>`__: Request: transformation functions - Yeo-Johnson
```
This and the authors list are generated during the release so you don't need to populate just now.

ilayn commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    ----------
+    x : ndarray
+        Input array.  Should be 1-dimensional.
+    lmbda : {None, scalar}, optional
```
`None` is implicit in the signature. `float, optional` is sufficient here.

ilayn commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    x : ndarray
+        Input array.  Should be 1-dimensional.
+    lmbda : {None, scalar}, optional
+        If `lmbda` is not None, do the transformation for that value.
```
These paragraphs can be combined. Note that you need double backticks for code formatting. 
```
If ``lmbda`` is None, find the lambda that maximizes the log-likelihood function and return it as the second output argument. Otherwise the transformation is done for the given value.
```

ilayn commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1253,337 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    -------
+    boxcox : ndarray
+        Box-Cox power transformed array.
+    maxlog : float, optional
```
We are explicitly avoiding varying number of outputs (cf. matlab). If it is cheap enough which seems to be the case here return it every time.

NicolasHug replied:
I think it's best to keep the API like this, as this is what boxcox is doing as well.

ilayn replied:
If it is a new function we don't want to increase the number of outliers but decrease them.

NicolasHug replied:
I can really see users experimenting with boxcox and yeojohnson in a generic manner, having a different signature complicates their code.

But OK.

I think that if lmbda is not None, the second returned value should be lmbda, or maybe None. Computing the best lambda every time is not that cheap:

```
In [2]: x =  stats.loggamma.rvs(5, size=1000000)                                                                                                                                               

In [4]: %timeit stats.yeojohnson(x, lmbda=None)                                                                                                                                                
2.16 s ± 1.93 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit stats.yeojohnson(x, lmbda=1.5)                                                                                                                                                 
71.6 ms ± 142 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

ilayn replied:
Hmm, let's call in some fresh eyes maybe I'm a bit too opinionated about this due to my matlab past.

@chrisb83 @josef-pkt @ev-br and others, do you mind weighing in for the variable number of output issue?

chrisb83 replied:
> I can really see users experimenting with boxcox and yeojohnson in a generic manner, having a different signature complicates their code.

I think that is a fair point, so I think keeping the API like it is for consistency can be justified.

ev-br replied:
The best would be probably to actually implement bunches and use them in both functions 
(original issue: https://github.com/scipy/scipy/issues/3665; pretty much PR ready class: https://github.com/scipy/scipy/pull/6335). But this could/should be done in a follow-up PR anyway.

Assuming we're not getting Bunches, I do suggest to think again about the optional returns.
(While I don't want to be a pain in the neck) All in all I'm -0 on optional returns.

While consistency with boxcox is indeed desirable, can it be achieved by always calling  `xt, _ = stats.boxcox(x)` and likewise for `yeojohnson`?

NicolasHug replied:
>While consistency with boxcox is indeed desirable, can it be achieved by always calling xt, _ = stats.boxcox(x) and likewise for yeojohnson?

If you only call `boxcox(x)` or `yeojohnson(x)` yes, but the problem is when you're passing a non-None `lmbda`:

```python
def yeojohnson_wrapper(x, lmbda):
    """Just a wrapper that makes yeo_johnson always return a tuple
    (proposed change)"""
    ret = stats.yeojohnson(x, lmbda)
    if not isinstance(ret, tuple):
        return ret, lmbda
    return ret

np.random.seed(54321)
x = stats.norm.rvs(size=100, loc=100)

for f in stats.yeojohnson, stats.boxcox:
    x_trans, lmbda_max = f(x, lmbda=None)  # works all the time
    x_trans = f(x, lmbda=3)  # works all the time

for f in yeojohnson_wrapper, stats.boxcox:
    x_trans, lmbda_max = f(x, lmbda=None)  # works for both
    x_trans, lmbda_max = f(x, lmbda=3)  # fails for boxcox that only return x_trans
```

To make it work, one would need to do something like this:
```python
for f in yeojohnson_wrapper, stats.boxcox:
    ret = f(x, lmbda=3)
    if f is yeojohnson_wrapper:
        x_trans = ret[0]
    else:
        x_trans = ret
```

I believe returning a Bunch is a great idea and would solve all problems.  Indeed it should probably be done in another PR.

ev-br replied:
Yup, this is convincing :-). Let's keep the API consistent in thus PR. 

A separate PR fir bunches in both functions would be good, but that's OT.

rgommers replied:
makes sense to me.

ilayn commented:
I've left some minor comments but in general I think you can reorder the examples such that the yeo-johnson code and plotting parts are separated to make it more  streamlined.

NicolasHug commented:
I addressed most of the comments, however I'm not sure what you mean about the examples?
Those were almost entirely copy/pasted from the boxcox examples.

chrisb83 commented:
Cross-ref bug report for boxcox in #6873 and proposed fix in #9271. Not sure if also relevant here.

NicolasHug commented:
Yes, I could reproduce the issue in #6873 with yeojohnson. We should probably wait for #9271 to be merged then?

NicolasHug commented on scipy/stats/morestats.py:
```diff
@@ -1215,28 +1254,332 @@ def boxcox_normplot(x, la, lb, plot=None, N=80):
+    Parameters
+    ----------
+    x : array_like
+        Input array.
```
Right, I just pushed a change so that `yeo_johnsonllf` now accepts multi-dimensional arrays and behaves like `boxcox_llf` so we can keep the docstring like that.

chrisb83 commented:
Looks like there are no major objections to keep the output consistent to `boxcox` (i.e. variable number of outputs)? I think this PR can then go into release 1.2.0.

rgommers commented:
> We still need a decision regarding output format. @rgommers, what is your view?

Looks like that's about the conversation that @ev-br resolved  where using Bunches is the ideal outcome (but for another) PR - commented on that, current status of this PR seems fine.

> I think this PR can then go into release 1.2.0.

+1

rgommers commented:
TravisCI test failures are real

NicolasHug commented:
I have no idea what's going on :/

The only difference between 6cd0370 (passing) and  dce2ea0 (failing) is a comment.

Also because the logs aren't synched: they show an error on line 1308 which is not the same as in my file.

I've merged master, without too much hope...

chrisb83 commented:
The tests fail when you check the inverse transformation. Results look good (i.e. close to expected result) but apparently do not pass the criterion.

Arrays are not almost equal to 2 decimals
E            ACTUAL: 0.0070733797505410281
E            DESIRED: 0

NicolasHug commented:
But why now? Those exact tests were passing before.

For some reason I can't install numpy 1.8.2 (I get a compilation error) so I can't reproduce the error for now

ilayn commented:
Let's try to get this one in before the 1.2 branch. 

@NicolasHug  Can you please switch to [`assert_allclose`](https://www.numpy.org/devdocs/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) anyways instead of `assert_equal`? In the past there were a lot of issues about it (didn't track the status). So maybe the old NumPy version is interfering

chrisb83 commented:
@NicolasHug, all: thanks, all checks are green, merged.

ilayn commented:
This one probably should go into the release notes too.

rgommers commented:
> This one probably should go into the release notes too.

Agreed. @NicolasHug would you mind adding a note to https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0?

NicolasHug commented:
Done :)

rgommers commented:
great, thanks
